contracts/src/utils/structs/bitmap.rs

```diff
@@ -0,0 +1,119 @@
+//! Contract module for managing uint256 to bool mapping in a compact and
+//! efficient way, provided the keys are sequential. Largely inspired by Uniswap's <https://github.com/Uniswap/merkle-distributor/blob/master/contracts/MerkleDistributor.sol>[merkle-distributor].
+//!
+//! `BitMap` packs 256 booleans across each bit of a single 256-bit slot of
+//! `uint256` type. Hence, booleans corresponding to 256 _sequential_ indices
+//! would only consume a single slot, unlike the regular `bool` which would
+//! consume an entire slot for a single value.
+//!
+//! This results in gas savings in two ways:
+//!
+//! - Setting a zero value to non-zero only once every 256 times
+//! - Accessing the same warm slot for every 256 _sequential_ indices
+use alloy_primitives::U256;
+use stylus_proc::sol_storage;
+
+sol_storage! {
+    /// State of bit map.
+    pub struct BitMap {
+        /// Inner laying mapping.
+        mapping(uint256 => uint256) _data;
+    }
+}
+
+impl BitMap {
+    /// Returns whether the bit at `index` is set.
+    ///
+    /// # Arguments
+    ///
+    /// * `index` - index of boolean value at the bit map.
+    #[must_use]
+    pub fn get(&self, index: U256) -> bool {
+        let bucket = index >> 8;
+        let mask = U256::from(1) << (index & U256::from(0xff));
+        let value = self._data.get(bucket);
+        (value & mask) != U256::ZERO
+    }
+
+    /// Sets the bit at `index` to the boolean `value`.
+    ///
+    /// # Arguments
+    ///
+    /// * `index` - index of boolean value at the bit map.
+    /// * `value` - boolean value to set into the bit map.
+    pub fn set_to(&mut self, index: U256, value: bool) {
+        if value {
+            self.set(index);
+        } else {
+            self.unset(index);
+        }
+    }
+
+    /// Sets the bit at `index`.
+    ///
+    /// # Arguments
+    ///
+    /// * `index` - index of boolean value that should be set `true`.
+    pub fn set(&mut self, index: U256) {
```

Comment:
Most of the code for `set` and `unset` is common, only last line differs. IMHO you can reuse the code.

Reply:
Agree. It was exactly like that at solidity
Added methods `get_mask` and `get_bucket` to reuse a bit of code